webapp/src/modules/bid/sagas.js

```diff
@@ -0,0 +1,164 @@
+  yield takeLatest(FETCH_BID_REQUEST, handleFetchBidById)
+  yield takeLatest(
+    FETCH_ASSET_ACCEPTED_BIDS_REQUEST,
+    handleFetchAssetAcceptedBidsRequest
+  )
+}
+
+function* handlePlaceBid({ bid }) {
+  try {
+    const { price, asset_type, asset_id, expires_at } = bid
+    const priceInWei = eth.utils.toWei(price)
+    const nftAddress = getNFTAddressByType(asset_type)
+    const estates = yield select(getEstates)
+    const asset = yield call(() => buildAsset(asset_id, asset_type, estates))
+    const bidder = yield select(getAddress)
+
+    let fingerprint = '0x'
+    if (asset_type === ASSET_TYPES.estate) {
+      const estateContract = eth.getContract('EstateRegistry')
+      fingerprint = yield call(() => estateContract.getFingerprint(asset.id))
+    }
+    const expiresIn = Math.round((expires_at - Date.now()) / 1000)
+
+    const bidContract = eth.getContract('ERC721Bid')
+
+    const txHash = yield call(() =>
+      bidContract.placeBid['address,uint256,uint256,uint256,bytes'](
+        nftAddress,
+        asset.id,
+        priceInWei,
+        expiresIn,
+        fingerprint
+      )
+    )
+    yield put(bidSuccess(txHash, { ...bid, bidder }, asset))
+    yield put(push(locations.activity()))
+  } catch (error) {
+    yield put(bidFailure(error.message))
+  }
+}
+
+function* handleCancelBid({ bid }) {
+  try {
+    const { token_address, token_id, asset_id, asset_type } = bid
+    const estates = yield select(getEstates)
+    const asset = yield call(() => buildAsset(asset_id, asset_type, estates))
+
+    const bidContract = eth.getContract('ERC721Bid')
+
+    const txHash = yield call(() =>
+      bidContract.cancelBid(token_address, token_id)
+    )
+    yield put(cancelBidSuccess(txHash, bid, asset))
+    yield put(push(locations.activity()))
+  } catch (error) {
+    yield put(cancelBidFailure(error.message))
+  }
+}
+
+function* handleAcceptBid({ bid }) {
+  try {
+    const { id, token_id, asset_id, asset_type } = bid
+    const erc721BidContractAddress = getContractAddress('ERC721Bid')
+    const estates = yield select(getEstates)
+    const asset = yield call(() => buildAsset(asset_id, asset_type, estates))
+    const owner = yield select(getAddress)
+
+    let contract
+    if (asset_type === ASSET_TYPES.parcel) {
+      contract = eth.getContract('LANDRegistry')
+    } else {
+      contract = eth.getContract('EstateRegistry')
+    }
+
+    const txHash = yield call(() =>
+      contract.safeTransferFrom['address,address,uint256,bytes'](
+        owner,
+        erc721BidContractAddress,
+        token_id,
+        id
+      )
+    )
+    yield put(acceptBidSuccess(txHash, { ...bid, seller: owner }, asset))
+    yield put(push(locations.activity()))
+  } catch (error) {
+    yield put(acceptBidFailure(error.message))
+  }
+}
+
+function* handleFetchBidById({ bidId }) {
+  try {
+    const bids = yield call(() => api.fetchBidById(bidId))
+    yield put(fetchBidByIdSuccess(bids[0]))
```

Comment:
Why does `fetchBidById` return an array? shouldn't it return a single bid?

Reply:
My bad